### ISSUE TITLE
Audinarey - Votes of expired token in gauges lead to dilution of FLOW emissions for gauges with active votes

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-06-velocimeter-judging#509]
Audinarey

High

# Votes of expired token in gauges lead to dilution of FLOW emissions for gauges with active votes

## Summary
If the user votes only once and does nothing afterwards, the voting power of the user does not decrease and this affects the overall rewards distribution result.

## Vulnerability Detail

The votingPower decreases linearly over time. At the point of voting, a users voting power is updated from the balance of the his `tokenId`. However, if no action is taken on the `tokenId` after the first vote and the token expires, the votingPower of that token at the point of voting remains in the `totalWeight` and `weights[_pool]`. This directly affects `_ratio` within the `Voter.notifyRewardAmount` function. It also impacts `claimable` in the `Voter._updateFor(...)` function.

```solidity
File: Voter.sol

249:     function _vote(uint _tokenId, address[] memory _poolVote, uint256[] memory _weights) internal {
250:         _reset(_tokenId);
251:         uint _poolCnt = _poolVote.length;
252:   >>>   uint256 _weight = IVotingEscrow(_ve).balanceOfNFT(_tokenId); // total weight of _tokenId in ve at the point of voting
....
256: 
257:         for (uint i = 0; i < _poolCnt; i++) {
258:             _totalVoteWeight += _weights[i]; // total/cummulative weight to add in the pool
259:         }
260: 
261:         for (uint i = 0; i < _poolCnt; i++) {
262:             address _pool = _poolVote[i];
263:             address _gauge = gauges[_pool];
264: 
265:             if (isGauge[_gauge]) {
...
272:                 poolVote[_tokenId].push(_pool);
273: 
274:      >>>        weights[_pool] += _poolWeight;
...
280:             }
281:         }
282:         if (_usedWeight > 0) IVotingEscrow(_ve).voting(_tokenId);
283:   >>>   totalWeight += uint256(_totalWeight);
284:         usedWeights[_tokenId] = uint256(_usedWeight);
285:     }


553:     function distribute(address _gauge) public lock { 
554:         IMinter(minter).update_period();
555:  >>>    _updateFor(_gauge); 
...
556:     }

521:     function _updateFor(address _gauge) internal { // it claims the reward for _gauge
522:         address _pool = poolForGauge[_gauge];
523:  >>>    uint256 _supplied = weights[_pool]; //  weight of votes in this pool
524:         if (_supplied > 0) {
525:             uint _supplyIndex = supplyIndex[_gauge]; // rewards amount for specific gauge tracker
...
538:     }

485:     function notifyRewardAmount(uint amount) external {
....
489:         _safeTransferFrom(base, msg.sender, address(this), amount); // transfer the distro in
490:  >>>    uint256 _ratio = amount * 1e18 / totalWeight; // 1e18 adjustment is removed during claim
...
495:     }


```

When the Voter.distribute function is called, newly emitted FLOW tokens are distributed to each gauge according to their weight contribution to the overall weight of all the gauges.

The problem is that, when  a token that was used to vote on a gauge expires, the weight of the token still contributes to the gauge's `weights[_pool]` and by extension to the `Voter.totalWeight`, hence A gauge will receive emissions for having dead weight in it pool diluting the FLOW emissions for pools that have votes with valid (unexpired) tokens.

- Alice and bob vote for gauge1 (G1) for one week with 500 weights each (G2 total weight = 1000)
- At the same time dan votes for gauge2 (G2) for 10 weeks with a weight of 1000
- Alice and bob lock have expired and are now dead weight in the G1's pool
- 4000e18 FLOW is emitted at the end of the first week and G1 and G2 receive 2000e18 each 
- at the end of the second epoch, 4000e18 FLOW is emitted and G1 and G2 receive 2000e18 each 
- the `tokenId`s used to votes in G1 have now expired but G1 still receives emissions equally as G2 with active votes
- G1 will continue to to receive emission despite having no active votes

## Impact

This leads to rewards dilution for gauges with active votes by gauges whose votes have expired.

## Code Snippet
https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/Voter.sol#L249-L251
https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/Voter.sol#L274
https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/Voter.sol#L283
https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/Voter.sol#L485-L490

## Tool used

Manual Review

## Recommendation
A trivial solution is not in sight because of possible side effects, but consider keeping tract of active votes in a gauge to distributes FLOW emissions instead the value of the weight that was supplied to the gauges poo at the time the user voted.

Duplicate of #171
